### ISSUE TITLE
Implement source code listing widget with line wrapping

### ISSUE THREAD

**Issue by strogonoff** [metanorma/metanorma.com#79]
Line-wrapping code is non-trivial: since newlines matter in source code, we have to differentiate between auto-wrapped lines and *actual* newlines. (This is why even source-code-heavy websites in the wild, like stackoverflow.com, rarely use this approach.)

I think the way to achieve this, UX-wise, is via line numbers which make it easy to tell the actual newlines visually.

Alternatively, wrapped newlines can be indented.

Neither of those approaches are doable with CSS from my research so far, but we’ll see. Likely this calls for a JS-based widget, which may be already available under suitable license (perhaps CodeMirror supports this), or it can be developed.

**Comment by ronaldtse** [comment]
Adding line numbers is a great suggestion.

Then maybe the accompanying approach is to break up long commands.

Let's not spend too much time in code widgets, just adopting a good enough one is sufficient.

**Comment by strogonoff** [comment]
Agreed about breaking longer commands, though I believe long lines may be unavoidable in some cases (either to demonstrate a certain feature or because of language limitations).

**Comment by strogonoff** [comment]
One upcoming change in a near release is new code listing font. Google Web Fonts offers good alternatives that are more readable and narrower than the default, letting more code fit in a line without making text really small.

**Comment by strogonoff** [comment]
Here’s an example of wrapping longer lines in command-line listings: https://github.com/metanorma/metanorma.com/commit/6090fc84764f39814bbbe505232bd116ce1bbab2

I’ll go over other content in a similar fashion in spare time, though I’ll mostly stay away from AsciiDoc examples since I am not 100% sure on how newlines behave there.

**Comment by opoudjis** [comment]
AsciiDoc: newlines do not matter in text in AsciiDoc. Newlines very much matter in commands  (which are any line not starting with a letter or number.)

**Comment by ronaldtse** [comment]
@opoudjis is right but AsciiDoc does cares about newlines in the document header (where there are attributes) though.

**Comment by doktorgee** [comment]
Implementing line numbers in lengthy code blocks makes sense. Maybe not as much for short examples. The one complication is that they would make copy-paste operations a bit inconvenient for anyone wanting to paste an example onto the command line (or possibly anywhere else).

**Comment by ronaldtse** [comment]
I’ve seen places where line numbers even when highlighted gets excluded from the copied text. I’m sure we can find an example of that?

**Comment by doktorgee** [comment]
+1 @ronaldtse, I'm sure I've seen that behavior too, somewhere.

**Comment by strogonoff** [comment]
> Implementing line numbers in lengthy code blocks makes sense. Maybe not as much for short examples. The one complication is that they would make copy-paste operations a bit inconvenient for anyone wanting to paste an example onto the command line (or possibly anywhere else).

If we want to wrap long lines, line numbers are crucial, otherwise it’s unclear where _actual_ new lines start.

Line numbers must not be selectable, of course—this requirement is what warrants the complexity of a widget (like CodeMirror) I mentioned in issue description in the first place.

**Comment by strogonoff** [comment]
I’ve seen examples of AsciiDoc markup where newlines matter, for example:

```
:very-long-document-attri
bute-name
```

This is not mentioning that sites deal with not just AsciiDoc but long shell commands and other languages. (Long shell command is what originally triggered Ronald’s concern.)

**Comment by strogonoff** [comment]
It should be noted that copy-pasting code, especially shell commands, from websites is an unsafe practice (it’s possible to produce a command that looks like string X but is actually completely different malicious string Y with the help of smart Unicode characters).

**Comment by strogonoff** [comment]
Closed accidentally.
